app/Jobs/ImportCsv.php

```diff
@@ -0,0 +1,119 @@
+<?php
+
+namespace App\Jobs;
+
+use Exception;
+use Throwable;
+use App\Models\User;
+use App\Models\Import;
+use App\Imports\Importer;
+use Carbon\CarbonInterface;
+use Illuminate\Support\Arr;
+use Illuminate\Bus\Batchable;
+use Illuminate\Bus\Queueable;
+use App\Models\FailedImportRow;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Queue\SerializesModels;
+use Illuminate\Queue\InteractsWithQueue;
+use Illuminate\Contracts\Queue\ShouldQueue;
+use Illuminate\Foundation\Bus\Dispatchable;
+use Illuminate\Validation\ValidationException;
+
+class ImportCsv implements ShouldQueue
+{
+    use Batchable;
+    use Dispatchable;
+    use InteractsWithQueue;
+    use Queueable;
+    use SerializesModels;
+
+    public bool $deleteWhenMissingModels = true;
+
+    protected readonly Importer $importer;
+
+    /**
+     * @param array<array<string, string>> $rows
+     * @param array<string, string> $columnMap
+     * @param array<string, mixed> $options
+     */
+    public function __construct(
+        readonly public Import $import,
+        readonly public array $rows,
+        readonly public array $columnMap,
+        readonly public array $options = [],
+    ) {
+        $this->importer = $this->import->getImporter(
+            $this->columnMap,
+            $this->options,
+        );
+    }
+
+    /**
+     * @return array<int, object>
+     */
+    public function middleware(): array
+    {
+        return $this->importer->getJobMiddleware();
+    }
+
+    public function handle(): void
+    {
+        /** @var User $user */
+        $user = $this->import->user;
+
+        auth()->login($user);
+
+        $exceptions = [];
+
+        foreach ($this->rows as $row) {
+            try {
+                DB::transaction(fn () => ($this->importer)($row));
+            } catch (ValidationException $exception) {
+                $this->logFailedRow($row, collect($exception->errors())->flatten()->implode(' '));
+            } catch (Throwable $exception) {
+                $exceptions[$exception::class] = $exception;
+
+                $this->logFailedRow($row);
+            }
+
+            $this->import->increment('processed_rows');
+        }
+
+        $this->handleExceptions($exceptions);
+    }
+
+    public function retryUntil(): CarbonInterface
+    {
+        return $this->importer->getJobRetryUntil();
+    }
+
+    /**
+     * @return array<int, string>
+     */
+    public function tags(): array
+    {
+        return $this->importer->getJobTags();
+    }
+
+    protected function logFailedRow(array $data, ?string $validationError = null): void
+    {
+        $failedRow = new FailedImportRow();
+        $failedRow->import()->associate($this->import);
+        $failedRow->data = $data;
+        $failedRow->validation_error = $validationError;
+        $failedRow->save();
```

Comment:
Can we use mass assignment here? Or is there a reason not to on `FailedImportRow`?

```suggestion
        $this->import->failedRows()->create([
            'data' => $data,
            'validation_error' => $validationError,
        ]);
```

Reply:
This will be moved to Filament at some point, and I don't particularly like mass assignment so this code will stay the same

Reply:
I am interested to hear why you don't like mass assignments that much. The suggestion seems a little more readable to me, but perhaps this is not always the case.

Reply:
It usually comes down to the fact that I am often associating multiple related models at once. With mass assignment, I need to then pass in `_id` columns into the array, which I don't like, as IMO the relationship should handle the specifics of the relationship columns with `associate()`.

I know that in this case we are only associating one relationship, but I like to keep things consistent, and only using mass assignment in some places and not others feels weird.

It's purely personal preference though.